### PR TITLE
fix: add and use post endpoint for streaming

### PR DESCRIPTION
**Pull request by NarekA**
Making a new PR that follows conventions to replace https://github.com/jina-ai/jina/pull/6091

Http streaming breaks when the input doc schema has fields which are not str, int, or float. This includes dicts, bools, and nested objects (See [Issue 6090](https://github.com/jina-ai/jina/issues/6090)). In addition it caps the input size to 2000 characters (Much less when you factor in URL encoding)

This PR changes the get endpoints to post endpoints. Not sure if we want to keep the get endpoints for backwards compatibility.

**Goals:**

- resolves [Issue 6090](https://github.com/jina-ai/jina/issues/6090)
- [x] check and update documentation. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#-contributing-documentation) and ask the team.

### CODE REVIEW

File: tests/integration/docarray_v2/test_issues.py [hunk 1]
```diff
@@ -92,3 +94,50 @@ def foo(self, docs: DocList[A], **kwargs) -> DocList[A]:
...
+        ):
+            docs.append(doc)
+
+    assert [d.text for d in docs] == [
+        "hello world my input text 0",
+        "hello world my input text 1",
+        "hello world my input text 2",
+        "hello world my input text 3",
+    ]
+    assert docs[0].nested_field.name == "test_name"
```
Comment (NarekA) [inline tests/integration/docarray_v2/test_issues.py:142]: @JoanFM I added this test to make sure complex Docs can be streamed

File: jina/serve/runtimes/gateway/http_fastapi_app_docarrayv2.py [hunk 2]
```diff
@@ -258,6 +258,29 @@ async def streaming_get(request: Request):
             async def event_generator():
                 async for doc, error in streamer.stream_doc(
                     doc=input_doc_model(**query_params), exec_endpoint=endpoint_path
+                ):
+                    if error:
+                        raise HTTPException(status_code=499, detail=str(error))
```
Comment (JoanFM) [inline jina/serve/runtimes/gateway/http_fastapi_app_docarrayv2.py:263]: can we get this 499 from the proper fastapi status constants?

File: tests/integration/docarray_v2/test_issues.py [hunk 3]
```diff
@@ -92,3 +94,50 @@ def foo(self, docs: DocList[A], **kwargs) -> DocList[A]:
...
+        @requests(on="/stream")
+        async def stream(
+            self, doc: InputWithComplexFields, parameters: Optional[Dict] = None, **kwargs
+        ) -> InputWithComplexFields:
+            for i in range(4):
+                yield InputWithComplexFields(text=f"hello world {doc.text} {i}")
+
+    docs = []
+    protocol = "http"
+    with Deployment(uses=MyExecutor, protocol=protocol, port=11112) as dep:
```
Comment (JoanFM) [inline tests/integration/docarray_v2/test_issues.py:125]: for the tests we have a helper method that gives us a random port, please use it instead of giving a fixed one. That method makes sure there are no collisions

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/serve/runtimes/worker/http_fastapi_app.py [hunk 4]
```diff
@@ -143,7 +143,27 @@ async def streaming_get(request: Request):
...
+        @app.api_route(
+            path=f'/{endpoint_path.strip("/")}',
+            methods=['POST'],
+            summary=f'Streaming Endpoint {endpoint_path}',
+        )
+        async def streaming_post(body: input_doc_model, request: Request):
+            req = DataRequest()
+            req.header.exec_endpoint = endpoint_path
+            if not docarray_v2:
+                from docarray import Document
```
Comment (JoanFM) [inline jina/serve/runtimes/worker/http_fastapi_app.py:161]: why is this import needed?

**Review by JoanFM — CHANGES_REQUESTED** [review]
I like how the solution looks, I just have some small comments.

File: jina/clients/base/helper.py [hunk 5]
```diff
@@ -197,19 +197,15 @@ async def send_streaming_message(self, doc: 'Document', on: str):
         :param on: Request endpoint
         :yields: responses
         """
-        if docarray_v2:
-            req_dict = doc.dict()
-        else:
-            req_dict = doc.to_dict()
+        req_dict = doc.json()
```
Comment (JoanFM) [inline jina/clients/base/helper.py:200]: I would not do this change, I think the signature of `json` may return a `str`. Why is there a need to change it?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/clients/base/helper.py (hunk 5 above)
Comment (NarekA) [reply jina/clients/base/helper.py:200]: In order to be passed as `request_kwargs['data']` rather than `request_kwargs['params']`, it needs to be serialized as a string.

File: jina/serve/runtimes/worker/http_fastapi_app.py (hunk 4 above)
Comment (NarekA) [reply jina/serve/runtimes/worker/http_fastapi_app.py:161]: Removed!

File: jina/serve/runtimes/gateway/http_fastapi_app_docarrayv2.py (hunk 2 above)
Comment (NarekA) [reply jina/serve/runtimes/gateway/http_fastapi_app_docarrayv2.py:263]: I can't seem to find a fastapi constant for this, can you find me an example? This code is copied from the existing get endpoint.

**Comment by NarekA** [comment]
@JoanFM I addressed the your comments. Please let me know If there's anything else I can do.

**Comment by JoanFM** [comment]
Hey @NarekA ,

Thanks for the contrubution, could yiu add a note in the documentation https://docs.jina.ai/concepts/serving/executor/add-endpoints/#streaming-endpoints making explicit note that SSE can only be used with flat schemas?

File: jina/clients/base/helper.py (hunk 5 above)
Comment (NarekA) [reply jina/clients/base/helper.py]: Was able to keep this by using `json` instead of `data`

File: docs/concepts/serving/executor/add-endpoints.md [hunk 6]
```diff
@@ -381,6 +381,11 @@ Streaming endpoints receive one Document as input and yields one Document at a t
 :class: note
 
 Streaming endpoints are only supported for HTTP and gRPC protocols and for Deployment and Flow with one single Executor.
+
+For HTTP deploymnent streaming executors generate both a GET and POST endpoint.
+The GET endpoint support documents with string, integer, or float fields only,
+whereas, POST requests support all docarrays.
+The Jina client uses the POST endpoints.
```
Comment (NarekA) [inline docs/concepts/serving/executor/add-endpoints.md:388]: @JoanFM Created the following docs

File: docs/concepts/serving/executor/add-endpoints.md [hunk 7]
```diff
@@ -381,6 +381,11 @@ Streaming endpoints receive one Document as input and yields one Document at a t
 :class: note
 
 Streaming endpoints are only supported for HTTP and gRPC protocols and for Deployment and Flow with one single Executor.
+
+For HTTP deploymnent streaming executors generate both a GET and POST endpoint.
```
Comment (JoanFM) [inline docs/concepts/serving/executor/add-endpoints.md:385]: typo in deployment

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: docs/concepts/serving/executor/add-endpoints.md (hunk 7 above)
Comment (NarekA) [reply docs/concepts/serving/executor/add-endpoints.md]: Fixed

File: jina/clients/base/helper.py (hunk 5 above)
Comment (JoanFM) [reply jina/clients/base/helper.py]: but where does the need for passing as 'str' come from?

Comment (NarekA) [reply jina/clients/base/helper.py]: If we were using `data`, data needs to be a string, but since I swithced to json, I am able to use dict again. Here is the current code:
```python
 'json': doc.dict() if docarray_v2 else doc.to_dict(),
```

**Comment by JoanFM** [comment]
> @JoanFM I addressed the your comments. Please let me know If there's anything else I can do.

It looks good to me, but there seems to be a failing test

**Comment by NarekA** [comment]
@JoanFM I am trying to fix the tests, but I'm having issues recreating CI tests locally, or even seeing what the failure is in CI.

File: tests/integration/docarray_v2/test_streaming.py [hunk 8]
```diff
@@ -143,17 +143,19 @@ async def test_streaming_delay(protocol, include_gateway):
...
-        async for doc in client.stream_doc(
+        stream = client.stream_doc(
             on='/hello',
             inputs=MyDocument(text='hello world', number=i),
             return_type=MyDocument,
-        ):
+        )
+        start_time = None
+        async for doc in stream:
+            start_time = start_time or time.time()
```
Comment (NarekA) [inline tests/integration/docarray_v2/test_streaming.py:153]: Looks like streaming takes longer to initialize, but has all of the data emitted at once:
```
Expected delay to be less than 0.5, got 7.867813110351562e-06 on iteration 1
Expected delay to be less than 1.0, got 0.00012993812561035156 on iteration 2
Expected delay to be less than 1.5, got 0.0001659393310546875 on iteration 3
Expected delay to be less than 2.0, got 0.00019884109497070312 on iteration 4
Expected delay to be less than 2.5, got 0.00022912025451660156 on iteration 5
``

Comment (NarekA) [reply tests/integration/docarray_v2/test_streaming.py:153]: All these times do not include a 2.5 second initialization time.

Comment (NarekA) [reply tests/integration/docarray_v2/test_streaming.py:153]: Just realizes the gateway seems to be breaking the streaming functionality here. Will look into why this is happening.

Comment (NarekA) [reply tests/integration/docarray_v2/test_streaming.py:153]: Looks like the gateway might be breaking streaming.

**Comment by JoanFM** [comment]
Thanks @NarekA for the great contribution
